apache/superset
# feat: Improve state key generation for dashboards and charts

### SUMMARY
When opening a dashboard or a chart in Explore, the URL will update with a new state key every time the page is reloaded even when the state isn't updated. This may not be scalable in high-traffic Superset deployments where it can quickly fill up disk space or exhaust the number of cache keys too fast, due to memory restrictions, making it more likely that a link you shared with someone a while ago becomes invalid.

To address this problem we should consider the request context such as user ID, resource ID, etc, to generate new keys only when the context changes. This way we can reuse existing keys but at the same time maintain the uniqueness of storage spaces for different users/dashboards/datasets/charts.

We could use HMAC to generate the same key for the same context but that would require key rotation management.

We can’t base key generation on previous keys because we have POST scenarios where the key is not present. One example would be when the user is inside a dashboard and selects “View chart in Explore”.

We can’t change the structure of the cache key to use the user and resource because this information is not present in the GET/DELETE requests and it would also require keeping backward compatibility with previous versions.

To implement the desired behavior we’ll keep a secondary mapping of user and resources with the previously generated keys. We'll store this mapping together with the entry so they’ll share the same expiration time. We’ll also keep it in sync with DELETE events. That way we can reuse the keys and keep the solution backward compatible.

Hat tip to @zhaoyongjie and @ktmud for bringing this concern during the reviews 👏🏼 

Follow-up of https://github.com/apache/superset/pull/18181

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
https://user-images.githubusercontent.com/70410625/152404373-a9680644-402c-4e48-a155-b47881f13993.mov

### TESTING INSTRUCTIONS
1 - Check that `form_data_key` is reused when the same user is accessing the same resource (dashboard, dataset, or chart)
2 - Check that `form_data_key` changes when the user, dashboard, dataset, or chart changes.
3 - Check that the values associated with a given key are correctly retrieved from the server

### ADDITIONAL INFORMATION
